### ISSUE TITLE
Catch all exceptions

### ISSUE THREAD

**Issue by ben221199** [laravel-json-api/laravel#62]
When I'm using Laravel, I have a seperate setup for API and Web. So, when having the API setup, all exceptions should be catched by the JSON API to return the error in JSON format.

At the moment, the Exception handler contains this code:
```PHP
public function register(){
	$this->renderable(ExceptionParser::make()->accept(static function(){
		return true;
	})->renderable());
}
```

This works, because the accept-callback returns `true`, and so every error will be converted.
Is there any way to implement the handler without the accept-callback?
The following codes don't work:
```PHP
public function register(){
	$this->renderable(ExceptionParser::make()->renderable());
}
```
and
```PHP
public function register(){
	$this->renderable(ExceptionParser::make()->append([
		HttpExceptionInterface::class,
	])->renderable());
}
```

There is also another project where the API and web are together in one setup. Is catching all exceptions also possible then? In that case it should only catch all exceptions in the `api` part.

BTW, when can I expect a stable version of this package?

**Comment by lindyhopchris** [comment]
@ben221199 sorry for being a bit slow to reply to this issue.

For the first scenario, are you effectively asking for this:

```php
$this->renderable(
    ExceptionParser::make()->alwaysAccept()->renderable()
);
```

I.e. add an `alwaysAccept()` method that means it will always render JSON:API?

For your example of an app that has both web and API routes, you'd need to add an `accept()` callback that determines whether or not the request was to the API or the web part of the application. That example is already shown in the documentation here:
https://laraveljsonapi.io/docs/1.0/responses/errors.html#custom-rendering-logic

In this example:

```php
$this->renderable(ExceptionParser::make()
    ->accept(fn(\Throwable $ex, $request) => $request->is('/api*'))
    ->renderable()
);
```

The following example you gave doesn't work:

```php
public function register(){
	$this->renderable(ExceptionParser::make()->append([
		HttpExceptionInterface::class,
	])->renderable());
}
```

Because the `append` and `prepend` methods are for attaching exception handlers, as per the docs here:
https://laraveljsonapi.io/docs/1.0/responses/errors.html#converting-exceptions

One thing you should be aware of is this exception parsing implementation is limited by what Laravel does in its exception handler. There are occassions where the handler doesn't even call any of the renderables registered by the `renderable()` method. As you can see here:
https://github.com/laravel/framework/blob/c351a309f1a02098f9a7ee24a8a402e9ce06fead/src/Illuminate/Foundation/Exceptions/Handler.php#L332

The renderables are not immediately called by the `render()` method, which means there will be some scenarios where they won't be invoked. As you can also see the renderables are called with the *parsed* exception, not the original exception - which also means there are some limitations as to whether an exception class will even hit our renderable implementation.

Anyway, if you could clarify what you need added that would help!

**Comment by ben221199** [comment]
That `alwaysAccept` function seems really nice. Adding that one would be cool.

The other option, when having Web and API on the same setup, isn't enough for me. I don't host my API under `/api`, but on another subdomain. I could check using domain instead of path, but I think there is a better way:

If you setup a plain Laravel installation, then there are two middlewares in the RouteServiceProvider: `web` and `api`. In my setup, `api` still exists. So, if you have a API under `/api` or on another subdomain, the both have `api`, so why don't make a function that checks on a specific middleware? For example `acceptMiddleware` or `acceptWithMiddleware`:

```php
$this->renderable(
	ExceptionParser::make()->acceptMiddleware('api')->renderable()
);
```

I also had a last question in my previous message: `BTW, when can I expect a stable version of this package?`
I would like to use this package on a stable setup, but I don't know when you will release it.

**Comment by lindyhopchris** [comment]
So this example:

```php
$this->renderable(ExceptionParser::make()
    ->accept(fn(\Throwable $ex, $request) => $request->is('/api*'))
    ->renderable()
);
```

Isn't saying that you *have* to use the `/api` path to detect whether the request is for the API. It's just showing an example of how you can use the `accept()` method - the logic for how you detect whether it's an API request would be down to your application. If you want to detect it from middleware, then that's fine - just use the `accept` method to do that. You might want to investigate what methods are available on the Laravel request.

If you want to add a method specifically for checking for the existence of middleware, then you'll need to submit a PR to the `laravel-json-api/exceptions` package.

I'm hoping to tag `1.0.0-rc.1` at the end of April or very early May. I've upgraded a production API to the new package on `1.0.0-beta.1` and it's all working fine. I haven't deployed that to production but will be doing so in late April/early May. So at that point I'll do `rc.1` which will be production ready. There's likely that a number of issues might still be open, but I'll either solve those during the `1.0.0-stable` series or defer them to `2.0.0`. Hope that helps!

**Comment by ben221199** [comment]
I understand your example of the `accept` function and how I should implement it myself using middleware. I only would like a specific accept for a specific middleware, because I think it would be useful for multiple purposes. I think I will make a PR for that.

End of April or early May would be cool. Thanks for the package.

**Comment by ben221199** [comment]
I made the pull request: https://github.com/laravel-json-api/exceptions/pull/1

**Comment by lindyhopchris** [comment]
Closing this as the new methods were tagged and released as `laravel-json-api/exceptions:1.0.0-beta.2`
